x-pack/plugins/licensing/public/plugin.test.ts

```diff
@@ -0,0 +1,296 @@
+        coreSetup.http.get.mockResolvedValue(fetchedLicense);
+
+        const { license$, refresh } = await plugin.setup(coreSetup);
+
+        refresh();
+        const license = await license$.pipe(take(1)).toPromise();
+
+        expect(license.uid).toBe('fresh');
+
+        expect(sessionStorage.setItem).toBeCalledTimes(1);
+
+        expect(sessionStorage.setItem.mock.calls[0][0]).toBe(licensingSessionStorageKey);
+        expect(sessionStorage.setItem.mock.calls[0][1]).toMatchInlineSnapshot(
+          `"{\\"license\\":{\\"uid\\":\\"fresh\\",\\"status\\":\\"active\\",\\"type\\":\\"basic\\",\\"expiryDateInMillis\\":5000},\\"features\\":{\\"ccr\\":{\\"enabled\\":true,\\"available\\":true},\\"ml\\":{\\"enabled\\":false,\\"available\\":true}},\\"signature\\":\\"xxxxxxxxx\\"}"`
+        );
+
+        const saved = JSON.parse(sessionStorage.setItem.mock.calls[0][1]);
+        expect(License.fromJSON(saved).toJSON()).toEqual(fetchedLicense.toJSON());
+      });
+
+      it('returns a license with error when request fails', async () => {
+        const sessionStorage = coreMock.createStorage();
+        plugin = new LicensingPlugin(coreMock.createPluginInitializerContext(), sessionStorage);
+
+        const coreSetup = coreMock.createSetup();
+        coreSetup.http.get.mockRejectedValue(new Error('reason'));
+
+        const { license$, refresh } = await plugin.setup(coreSetup);
+        refresh();
+
+        const license = await license$.pipe(take(1)).toPromise();
+
+        expect(license.isAvailable).toBe(false);
+        expect(license.error).toBeDefined();
+        expect(license.error!.message).toBe('reason');
+      });
+
+      it('remove license saved in session storage when request failed', async () => {
+        const sessionStorage = coreMock.createStorage();
+        plugin = new LicensingPlugin(coreMock.createPluginInitializerContext(), sessionStorage);
+
+        const coreSetup = coreMock.createSetup();
+        coreSetup.http.get.mockRejectedValue(new Error('sorry'));
+
+        const { license$, refresh } = await plugin.setup(coreSetup);
+        expect(sessionStorage.removeItem).toHaveBeenCalledTimes(0);
+
+        refresh();
+        await license$.pipe(take(1)).toPromise();
+
+        expect(sessionStorage.removeItem).toHaveBeenCalledTimes(1);
+        expect(sessionStorage.removeItem).toHaveBeenCalledWith(licensingSessionStorageKey);
+      });
+    });
+  });
+  describe('interceptor', () => {
+    it('register http interceptor checking signature header', async () => {
+      const sessionStorage = coreMock.createStorage();
+      plugin = new LicensingPlugin(coreMock.createPluginInitializerContext(), sessionStorage);
+
+      const coreSetup = coreMock.createSetup();
+
+      await plugin.setup(coreSetup);
+      expect(coreSetup.http.intercept).toHaveBeenCalledTimes(1);
+    });
+
+    it('http interceptor triggers re-fetch if signature header has changed', async () => {
```

Comment:
The interceptor tests here and below would probably be a lot less brittle and simpler if the interceptor function was exported on it's own so it could be tested without having to do this dance with mocks to get it out.